File: src/harpy/lexer/lexer.py
from .token import Token
from .token_type import TokenType


class Lexer:
    """A very primitive lexer. Takes a string and splits it into a series of
    Tokens. Operators and punctuation are mapped to unique keywords. Names,
    which can be any series of letters, are turned into NAME tokens. All other
    characters are ignored (except to separate names). Numbers and strings are
    not supported. This is really just the bare minimum to give the parser
    something to work with.
    """

    _index: int
    _text: str
    _directives: dict[str, TokenType]
    _keywords: dict[str, TokenType]
    _simple_operators: dict[str, TokenType]
    _compound_operatos: dict[str, TokenType]
    _names: list[str]
    _line: int
    _pos: int
    _reset_index: int

    def __init__(self, text: str):
        """Creates a new Lexer to tokenize the given string.

        Args:
            text (str): String to tokenize.
        """

        self._text = text
        self._directives = {}
        self._keywords = {}
        self._compound_operators = {}
        self._simple_operators = {}
        self._names = []
        self._line = 1
        self._index = 0
        self._pos = 0
        self._reset_index = 0

        for type in TokenType:
            directive = type.preprocessor_directive()
            if directive is not None:
                self._directives[directive] = type
                continue

            keyword = type.keyword()
            if keyword is not None:
                self._keywords[keyword] = type
                continue

            compound_operator = type.compound_operator()
            if compound_operator is not None:
                self._compound_operators[compound_operator] = type
                continue

            simple_operator = type.simple_operator()
            if simple_operator is not None:
                self._simple_operators[simple_operator] = type
                continue

    def __iter__(self):
        return self

    def __next__(self):
        while self._index < len(self._text):
            c = self._advance()

            match c:
                case "#":
                    return self._read_preprocessor_directive()
                case "/":
                    # Potential comments.
                    match self._peek():
                        case "/":
                            return self._read_line_comment()
                        case "*":
                            return self._read_block_comment()
                        case _:
                            return Token(
                                type=self._simple_operators[c],
                                text=c,
                                line=self._line,
                                position=self._pos,
                            )
                case "[":
                    if self._peek().isalpha():
                        return self._read_str_literal_or_bracket(c)
                    else:
                        return Token(
                            type=self._simple_operators[c],
                            text=c,
                            line=self._line,
                            position=self._pos,
                        )
                case '"' | "'":
                    return self._read_str_literal_or_bracket(c)
                case ".":
                    if self._peek().isalpha():
                        return self._read_bool_literal_or_logical()
                    else:
                        return self._read_num_literal(c)
                case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9":
                    return self._read_num_literal(c)
                case _:
                    if (kw := self._read_keyword(c)) is not None:
                        return kw
                    elif c + (c1 := self._peek()) in self._compound_operators:
                        self._advance()
                        return Token(
                            type=self._compound_operators[c + c1],
                            text=c + c1,
                            line=self._line,
                            position=self._pos,
                        )
                    elif c in self._simple_operators:
                        return Token(
                            type=self._simple_operators[c],
                            text=c,
                            line=self._line,
                            position=self._pos,
                        )
                    elif c.isalpha():
                        return self._read_name()
                    else:
                        if c == "\n":
                            self._line += 1
                            self._pos = 0
                        # Ignore all other characters (whitespace, etc.)
                        pass

        # Once we've reached the end of the string, just return EOF tokens. We'll
        # just keep returning them as many times as we're asked so that the
        # parser's lookahead doesn't have to worry about running out of tokens.
        return Token(type=TokenType.EOF, text="\0", line=self._line, position=self._pos)

    def _read_preprocessor_directive(self) -> Token:
        start_index = self._index - 1
        self._set_reset_index(index=start_index)
        directive = ""

        while (c := self._advance()).isalpha():
            directive += c

        if directive.lower() in self._directives:
            while True:
                match self._peek():
                    case "\n" | "\r" | "\0":
                        return Token(
                            type=self._directives[directive.lower()],
                            text=self._text[start_index : self._index],
                            line=self._line,
                            position=self._pos,
                        )
                    case _:
                        self._advance()

        self._reset(offset=1)

        return Token(type=TokenType.NE1, text="#", line=self._line, position=self._pos)

    def _read_line_comment(self) -> Token:
        start_index = self._index - 1

        # Consume second '/'.
        self._advance()

        while True:
            match self._peek():
                case "\n" | "\r" | "\0":
                    return Token(
                        type=TokenType.LINE_COMMENT,
                        text=self._text[start_index : self._index],
                        line=self._line,
                        position=self._pos,
                    )
                case _:
                    self._advance()

    def _read_block_comment(self) -> Token:
        start_index = self._index - 1

        while True:
            match c := self._advance():
                case "*":
                    match self._advance():
                        case "/":
                            return Token(
                                type=TokenType.BLOCK_COMMENT,
                                text=self._text[start_index : self._index],
                                line=self._line,
                                position=self._pos,
                            )
                        case "\0":
                            raise SyntaxError("Unterminated block comment.")
                        case _:
                            if c == "\n":
                                self._pos = 0
                            pass  # Do nothing, keep advancing.
                case "\0":
                    raise SyntaxError("Unterminated block comment.")
                case _:
                    if c == "\n":
                        self._pos = 0
                    pass  # Do nothing, keep advancing.

    def _read_bool_literal_or_logical(self) -> Token:
        literal = "."

        while (c := self._advance()) not in (".", "\0"):
            literal += c

        literal += "."

        match literal.lower():
            case ".t." | ".f.":
                return Token(
                    type=TokenType.BOOL_LITERAL,
                    text=literal,
                    line=self._line,
                    position=self._pos,
                )
            case ".or.":
                return Token(
                    type=TokenType.OR,
                    text=literal,
                    line=self._line,
                    position=self._pos,
                )
            case ".and.":
                return Token(
                    type=TokenType.AND,
                    text=literal,
                    line=self._line,
                    position=self._pos,
                )
            case _:
                raise SyntaxError(f"Unable to read token '{literal}'.")

    def _read_num_literal(self, c: str) -> Token:
        literal = c
        dotfound = False
        hexnum = False

        while not (c := self._peek()).isspace():
            match c.lower():
                case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9":
                    literal += c
                    self._advance()
                case "x":
                    if literal == "0":
                        literal += c
                        hexnum = True
                        self._advance()
                    else:
                        raise SyntaxError(
                            f"Unterminated hexadecimal literal '{literal + c}'."
                        )
                case "a" | "b" | "c" | "d" | "e" | "f":
                    if hexnum:
                        literal += c
                        self._advance()
                    else:
                        raise SyntaxError(f"Invalid numeric literal '{literal + c}'.")
                case ".":
                    if not dotfound:
                        literal += c
                        dotfound = True
                        self._advance()
                    else:
                        raise SyntaxError(
                            f"Second decimal point found in literal '{literal + c}'."
                        )
                case "\0":
                    return Token(
                        type=TokenType.NUM_LITERAL,
                        text=literal,
                        line=self._line,
                        position=self._pos,
                    )
                case _:
                    return Token(
                        type=TokenType.NUM_LITERAL,
                        text=literal,
                        line=self._line,
                        position=self._pos,
                    )

        return Token(
            type=TokenType.NUM_LITERAL,
            text=literal,
            line=self._line,
            position=self._pos,
        )

    def _read_str_literal_or_bracket(self, c: str) -> Token:
        start_index = self._index
        self._set_reset_index(index=start_index)
        literal = c
        endquote = "]" if c == "[" else c

        while (c := self._peek()) != endquote:
            match c:
                case "\0":
                    if len(literal) > 1 and literal[-1:] == endquote:
                        break
                    raise SyntaxError(f"Unterminated string literal '{literal}'.")
                case _:
                    literal += c

            self._advance()

        if endquote == "]" and literal[1:] in self._names:
            self._reset()
            return Token(
                type=TokenType.LEFT_BRACKET,
                text="[",
                line=self._line,
                position=self._pos,
            )

        return Token(
            type=TokenType.STR_LITERAL,
            text=literal + self._advance(),
            line=self._line,
            position=self._pos,
        )

    def _read_keyword(self, kw: str) -> Token | None:
        self._set_reset_index(index=self._index)

        while (c := self._peek()).isalpha():
            kw += c
            self._advance()

        if kw.lower() in self._keywords:
            return Token(
                type=self._keywords[kw.lower()],
                text=kw,
                line=self._line,
                position=self._pos,
            )

        self._reset()
        return None

    def _read_name(self) -> Token:
        start_index = self._index - 1
        self._set_reset_index(index=start_index)

        while self._peek() != "\0":
            if not self._text[self._index].isalnum() and self._text[self._index] != "_":
                break
            self._advance()

        name = self._text[start_index : self._index]

        self._names.append(name)

        return Token(
            type=TokenType.NAME, text=name, line=self._line, position=self._pos
        )

    def _peek(self) -> str:
        if len(self._text) > self._index:
            return self._text[self._index]
        else:
            return "\0"

    def _advance(self) -> str:
        c = "\0"

        if len(self._text) > self._index:
            c = self._text[self._index]
            if len(self._text) > self._index:
                self._index += 1
                self._pos += 1

        return c

    def _set_reset_index(self, index: int):
        self._reset_index = index

    def _reset(self, offset: int = 0):
        self._pos -= self._index - self._reset_index + offset
        if (
            self._pos < 0
        ):  # Be careful not to reset further than the same line, since the line number will now be incorrect
            self._pos = 0
        self._index = self._reset_index + offset
        self._reset_index = 0

File: Harpy/Lexer/Lexer.cs
namespace Harpy.Lexer;

/// <summary>
///     Takes a string and splits it into a series of <see cref="HarbourSyntaxToken" />s.
///     Operators and punctuation are mapped to unique keywords.
/// </summary>
public class Lexer(string text)
{
    private readonly List<string> _names = [];

    private int _index;
    private int _line = 1;
    private int _pos;
    private int _resetIndex;

    public IEnumerator<HarbourSyntaxToken> GetEnumerator()
    {
        List<HarbourSyntaxTrivia> leadingTrivia = [];
        List<HarbourSyntaxTrivia> trailingTrivia = [];
        HarbourSyntaxToken? oldToken = null;
        HarbourSyntaxToken? newToken = null;
        var newLineEncountered = false;

        while (_index < text.Length)
        {
            var c = Advance();

            switch (c)
            {
                case '#':
                {
                    if (newLineEncountered)
                        leadingTrivia.Add((HarbourSyntaxTrivia)ReadPreprocessorDirectiveOrNeOp());
                    else
                        trailingTrivia.Add((HarbourSyntaxTrivia)ReadPreprocessorDirectiveOrNeOp());

                    break;
                }
                case '/':
                {
                    switch (Peek())
                    {
                        case '/':
                        {
                            if (newLineEncountered)
                                leadingTrivia.Add(ReadLineComment());
                            else
                                trailingTrivia.Add(ReadLineComment());
                            break;
                        }
                        case '*':
                        {
                            if (newLineEncountered)
                                leadingTrivia.Add(ReadBlockComment());
                            else
                                trailingTrivia.Add(ReadBlockComment());
                            break;
                        }
                        default:
                        {
                            newToken = new HarbourSyntaxToken(
                                HarbourSyntaxToken.SimpleOperatorFromText(c.ToString()),
                                c.ToString(),
                                _line,
                                _pos
                            );
                            break;
                        }
                    }

                    break;
                }
                case '[':
                {
                    newToken = char.IsLetterOrDigit(Peek())
                        ? ReadStringLiteralOrBrackets(c)
                        : new HarbourSyntaxToken(HarbourSyntaxKind.LEFT_BRACKET, c.ToString(), _line, _pos);

                    break;
                }
                case '"':
                case '\'':
                {
                    newToken = ReadStringLiteralOrBrackets(c);
                    break;
                }
                case '.':
                {
                    newToken = char.IsLetter(Peek()) ? ReadBooleanLiteralOrLogical(c) : ReadNumericLiteral(c);
                    break;
                }
                case '0':
                case '1':
                case '2':
                case '3':
                case '4':
                case '5':
                case '6':
                case '7':
                case '8':
                case '9':
                {
                    newToken = ReadNumericLiteral(c);
                    break;
                }
                default:
                {
                    var keyword = ReadKeyword(c);
                    var op = c.ToString() + Peek();

                    if (keyword != null)
                    {
                        newToken = keyword;
                    }
                    else if (HarbourSyntaxToken.CompoundOperators.ContainsValue(op))
                    {
                        Advance();
                        newToken = new HarbourSyntaxToken(
                            HarbourSyntaxToken.CompoundOperatorFromText(op),
                            op,
                            _line,
                            _pos
                        );
                    }
                    else if (HarbourSyntaxToken.SimpleOperators.ContainsValue(c.ToString()))
                    {
                        newToken = new HarbourSyntaxToken(
                            HarbourSyntaxToken.SimpleOperatorFromText(c.ToString()),
                            c.ToString(),
                            _line,
                            _pos
                        );
                    }
                    else if (char.IsLetterOrDigit(c))
                    {
                        newToken = ReadName();
                    }
                    else if (char.IsWhiteSpace(c) || c == ';')
                    {
                        HarbourSyntaxKind type;
                        if (op == "\r\n")
                        {
                            Advance();
                            type = HarbourSyntaxKind.NEWLINE;
                        }
                        else
                        {
                            type = c switch
                            {
                                '\n' => HarbourSyntaxKind.NEWLINE,
                                '\r' => HarbourSyntaxKind.CARRIAGE_RETURN,
                                '\t' => HarbourSyntaxKind.TAB,
                                ' ' => HarbourSyntaxKind.SPACE,
                                ';' => HarbourSyntaxKind.LINE_CONTINUATION,
                                _ => throw new InvalidSyntaxException($"Unknown character '{c}' encountered in source.")
                            };
                        }

                        if (newLineEncountered)
                            leadingTrivia.Add(
                                new HarbourSyntaxTrivia(
                                    type,
                                    c.ToString(),
                                    _line,
                                    _pos
                                )
                            );
                        else
                            trailingTrivia.Add(
                                new HarbourSyntaxTrivia(
                                    type,
                                    c.ToString(),
                                    _line,
                                    _pos
                                )
                            );

                        if (type is HarbourSyntaxKind.NEWLINE or HarbourSyntaxKind.CARRIAGE_RETURN)
                        {
                            _line += 1;
                            _pos = 0;
                            newLineEncountered = true;
                        }
                    }

                    break;
                }
            }

            if (newToken == null) continue; // Search for more trivia.
            newLineEncountered = false;
            newToken.LeadingTrivia = leadingTrivia;
            leadingTrivia = [];

            if (oldToken != null)
            {
                oldToken.TrailingTrivia = trailingTrivia;
                trailingTrivia = [];
                yield return oldToken;
            }

            oldToken = newToken;
            newToken = null;
        }

        if (oldToken != null)
        {
            oldToken.TrailingTrivia = trailingTrivia;
            yield return oldToken;
        }

        // Once we've reached the end of the string, just return EOF tokens. We'll
        // just keep returning them as many times as we're asked so that the
        // parser's lookahead doesn't have to worry about running out of tokens.
        leadingTrivia.AddRange(trailingTrivia);
        yield return new HarbourSyntaxToken(HarbourSyntaxKind.EOF, "\0", _line, _pos, leadingTrivia, []);
    }


    private HarbourSyntaxElement ReadPreprocessorDirectiveOrNeOp()
    {
        var startIndex = _index - 1;
        SetResetIndex(startIndex);
        var directive = "";
        char c;

        while (char.IsLetterOrDigit(c = Advance())) directive += c;

        if (HarbourSyntaxTrivia.Directives.ContainsValue(directive.ToLower()))
            while (true)
                switch (Peek())
                {
                    case '\n':
                    case '\r':
                    case '\0':
                    {
                        return new HarbourSyntaxTrivia(
                            HarbourSyntaxTrivia.DirectiveFromText(directive.ToLower()),
                            text[startIndex.._index],
                            _line,
                            _pos
                        );
                    }
                    default:
                    {
                        Advance();
                        break;
                    }
                }

        ResetIndex(1);

        return new HarbourSyntaxToken(HarbourSyntaxKind.NE1, "#", _line, _pos);
    }

    private HarbourSyntaxTrivia ReadLineComment()
    {
        var startIndex = _index - 1;

        Advance();

        while (true)
            switch (Peek())
            {
                case '\n':
                case '\r':
                case '\0':
                {
                    return new HarbourSyntaxTrivia(
                        HarbourSyntaxKind.LINE_COMMENT,
                        text[startIndex.._index],
                        _line,
                        _pos
                    );
                }
                default:
                {
                    Advance();
                    break;
                }
            }
    }

    private HarbourSyntaxTrivia ReadBlockComment()
    {
        var startIndex = _index - 1;
        var startLine = _line;

        while (true)
        {
            var c = Advance();

            switch (c)
            {
                case '*':
                {
                    var c1 = Advance();

                    switch (c1)
                    {
                        case '/':
                        {
                            return new HarbourSyntaxTrivia(
                                HarbourSyntaxKind.BLOCK_COMMENT,
                                text[startIndex.._index],
                                startLine,
                                _pos
                            );
                        }
                        case '\0':
                        {
                            throw new InvalidSyntaxException(
                                $"Unterminated block comment starting at line {startLine}.");
                        }
                        default:
                        {
                            if (c1 == '\n')
                            {
                                _line += 1;
                                _pos = 0;
                            }

                            break;
                        }
                    }

                    break;
                }
                case '\0':
                {
                    throw new InvalidSyntaxException($"Unterminated block comment starting at line {startLine}.");
                }
                default:
                {
                    if (c == '\n')
                    {
                        _line += 1;
                        _pos = 0;
                    }

                    break;
                }
            }
        }
    }

    private HarbourSyntaxToken ReadBooleanLiteralOrLogical(char c)
    {
        var literal = c.ToString();

        while ((c = Advance()) != '.' && c != '\0') literal += c;

        literal += '.';

        return literal.ToLower() switch
        {
            ".t." or ".f." => new HarbourSyntaxToken(HarbourSyntaxKind.BOOL_LITERAL, literal, _line, _pos),
            ".or." => new HarbourSyntaxToken(HarbourSyntaxKind.OR, literal, _line, _pos),
            ".and." => new HarbourSyntaxToken(HarbourSyntaxKind.AND, literal, _line, _pos),
            _ => throw new InvalidSyntaxException($"Unable to read token '{literal}'.")
        };
    }

    private HarbourSyntaxToken ReadNumericLiteral(char c)
    {
        var literal = c.ToString();
        var dotFound = false;
        var hexNum = false;

        c = Peek();
        while (!char.IsWhiteSpace(c))
        {
            switch (char.ToLower(c))
            {
                case '0':
                case '1':
                case '2':
                case '3':
                case '4':
                case '5':
                case '6':
                case '7':
                case '8':
                case '9':
                {
                    literal += c;
                    break;
                }
                case 'x':
                {
                    if (literal != "0")
                        throw new InvalidSyntaxException($"Unterminated hexadecimal literal '{literal + c}'.");
                    literal += c;
                    hexNum = true;
                    break;
                }
                case 'a':
                case 'b':
                case 'c':
                case 'd':
                case 'e':
                case 'f':
                {
                    if (!hexNum) throw new InvalidSyntaxException($"Invalid numeric literal '{literal + c}'.");
                    literal += c;
                    break;
                }
                case '.':
                {
                    if (dotFound)
                        throw new InvalidSyntaxException($"Second decimal point found in literal '{literal + c}'.");
                    literal += c;
                    dotFound = true;
                    break;
                }
                default:
                {
                    return new HarbourSyntaxToken(
                        HarbourSyntaxKind.NUM_LITERAL,
                        literal,
                        _line,
                        _pos
                    );
                }
            }

            Advance();

            c = Peek();
        }

        return new HarbourSyntaxToken(
            HarbourSyntaxKind.NUM_LITERAL,
            literal,
            _line,
            _pos
        );
    }

    private HarbourSyntaxToken ReadStringLiteralOrBrackets(char c)
    {
        var startIndex = _index;
        var literal = c.ToString();
        var endQuote = c == '[' ? ']' : c;

        SetResetIndex(startIndex);

        while ((c = Peek()) != endQuote)
        {
            switch (c)
            {
                case '\0':
                {
                    if (literal.Length > 1 && literal.EndsWith(endQuote)) break;
                    throw new InvalidSyntaxException($"Unterminated string literal '{literal}'.");
                }
                default:
                {
                    literal += c;
                    break;
                }
            }

            Advance();
        }

        // TODO: This won't work unless we handle include files/preprocessor parsing first,
        // e.g., `IF !(aHWFlag[F_HWCardReader] .OR. aHWFlag[F_HWDallasKey]) .OR. oPOSStatus:oFree:lFlag3`
        if (endQuote != ']' || !_names.Contains(literal[1..]))
            return new HarbourSyntaxToken(
                HarbourSyntaxKind.STR_LITERAL,
                literal + Advance(),
                _line,
                _pos
            );
        ResetIndex();
        return new HarbourSyntaxToken(HarbourSyntaxKind.LEFT_BRACKET, "[", _line, _pos);
    }

    private HarbourSyntaxToken? ReadKeyword(char c)
    {
        var keyword = c.ToString();
        SetResetIndex(_index);

        c = Peek();
        while (char.IsLetterOrDigit(c))
        {
            keyword += c;
            Advance();
            c = Peek();
        }

        if (HarbourSyntaxToken.Keywords.ContainsValue(keyword.ToLower()))
            return new HarbourSyntaxToken(
                HarbourSyntaxToken.KeywordFromText(keyword.ToLower()),
                keyword,
                _line,
                _pos
            );

        ResetIndex();
        return null;
    }

    private HarbourSyntaxToken ReadName()
    {
        var startIndex = _index - 1;
        SetResetIndex(startIndex);

        while (Peek() != '\0')
        {
            if (!char.IsLetterOrDigit(text[_index]) && text[_index] != '_') break;

            Advance();
        }

        var name = text[startIndex.._index];

        _names.Add(name);

        return new HarbourSyntaxToken(
            HarbourSyntaxKind.NAME,
            name,
            _line,
            _pos
        );
    }

    private char Peek()
    {
        return text.Length > _index ? text[_index] : '\0';
    }

    private char Advance()
    {
        var c = '\0';

        if (text.Length <= _index) return c;
        c = text[_index];
        _index += 1;
        _pos += 1;

        return c;
    }

    private void SetResetIndex(int index)
    {
        _resetIndex = index;
    }

    private void ResetIndex(int offset = 0)
    {
        _pos -= _index - _resetIndex + offset;
        if (_pos < 0)
            // Be careful not to reset further than the same line, since the line number will now be incorrect.
            _pos = 0;

        _index = _resetIndex + offset;
        _resetIndex = 0;
    }
}